[Linter] Add linter to detect device-bias hard code in test cases.

etaf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152948

Since XPU does not gate community pull requests, we’ve observed that contributors often hardcode "cuda" in functions decorated with @requires_gpu() when adding new test cases. This causes the tests to fail on XPU and breaks XPU CI.
This PR adds a linter to detect such issues automatically. An example is shown below.

```
  Error (TEST_DEVICE_BIAS) [device-bias]
    `@requires_gpu` function should not hardcode device='cuda'

        11670  |                .contiguous()
        11671  |            )
        11672  |
    >>> 11673  |        inp = torch.rand((64, 64), device="cuda") * 2 - 1
        11674  |        boundaries = torch.tensor([-0.9, -0.8, 0.1, 0.2, 0.5, 0.9])
        11675  |
        11676  |        self.common(fn, (inp, boundaries), check_lowp=False)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `@requires_gpu` function should not hardcode .cuda() call

        11700  |            self.assertEqual(ref, res)
        11701  |
        11702  |            for offset2 in (0, 1, 2, 3, 4):
    >>> 11703  |                base2 = torch.randn(64 * 64 + 64, dtype=torch.float32).cuda()
        11704  |                inp2 = torch.as_strided(base2, (64, 64), (64, 1), offset2)
        11705  |                ref2 = fn(inp2)
        11706  |                res2 = fn_c(inp2)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `@requires_gpu` function should not hardcode torch.device('cuda:0')

        11723  |            return x.sin() + x.cos()
        11724  |
        11725  |        base = torch.randn(
    >>> 11726  |            64 * 64 + 64, dtype=torch.float32, device=torch.device("cuda:0")
        11727  |        )
        11728  |
        11729  |        inp1 = torch.as_strided(base, (32, 32), (32, 1), 4)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `@requires_gpu` function should not hardcode .to('cuda') call

        11771  |            torch.manual_seed(42)
        11772  |            base = torch.randn(64 * 64 + 64, dtype=torch.float32, device=self.device)
        11773  |            torch.manual_seed(42)
    >>> 11774  |            base_ref = torch.randn(64 * 64 + 64, dtype=torch.float32).to("cuda")
        11775  |
        11776  |            inp = torch.as_strided(base, size, stride, offset)
        11777  |            inp_ref = torch.as_strided(base_ref, size, stride, offset)
```

etaf committed:
[Linter] Add linter to detect device-bias hard code in test cases.

[ghstack-poisoned]

Skylion007 commented on tools/linter/adapters/test_device_bias_linter.py:
```diff
@@ -0,0 +1,140 @@
+
+    args = parser.parse_args()
+
+    pool = mp.Pool(8)
+    lint_messages = pool.map(check_file, args.filenames)
+    pool.close()
+    pool.join()
```
```suggestion
    with mp.Pool(8) as pool:
        lint_messages = pool.map(check_file, args.filenames)
```

etaf replied:
Updated, thanks for your review.

etaf committed:
Update on "[WIP][Linter] Add linter to detect device-bias hard code in test cases."


Since XPU does not gate community pull requests, we’ve observed that contributors often hardcode "cuda" in functions decorated with requires_gpu() when adding new test cases. This causes the tests to fail on XPU and breaks XPU CI.
This PR adds a linter to detect such issues automatically. An example is shown below.

```
  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu()` function should not hardcode torch.device('cuda')

        11670  |                .contiguous()
        11671  |            )
        11672  |
    >>> 11673  |        inp = torch.rand((64, 64), device=torch.device("cuda")) * 2 - 1
        11674  |        boundaries = torch.tensor([-0.9, -0.8, 0.1, 0.2, 0.5, 0.9])
        11675  |
        11676  |        self.common(fn, (inp, boundaries), check_lowp=False)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu()` function should not hardcode .cuda() call

        11700  |            self.assertEqual(ref, res)
        11701  |
        11702  |            for offset2 in (0, 1, 2, 3, 4):
    >>> 11703  |                base2 = torch.randn(
        11704  |                    64 * 64 + 64, dtype=torch.float32, device=self.device
        11705  |                ).cuda()
        11706  |                inp2 = torch.as_strided(base2, (64, 64), (64, 1), offset2)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu()` function should not hardcode torch.device('cuda')

        11725  |            return x.sin() + x.cos()
        11726  |
        11727  |        base = torch.randn(
    >>> 11728  |            64 * 64 + 64, dtype=torch.float32, device=torch.device("cuda")
        11729  |        )
        11730  |
        11731  |        inp1 = torch.as_strided(base, (32, 32), (32, 1), 4)
```


[ghstack-poisoned]

etaf committed:
Update on "[Linter] Add linter to detect device-bias hard code in test cases."


Since XPU does not gate community pull requests, we’ve observed that contributors often hardcode "cuda" in functions decorated with requires_gpu() when adding new test cases. This causes the tests to fail on XPU and breaks XPU CI.
This PR adds a linter to detect such issues automatically. An example is shown below.

```
  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu()` function should not hardcode torch.device('cuda')

        11670  |                .contiguous()
        11671  |            )
        11672  |
    >>> 11673  |        inp = torch.rand((64, 64), device=torch.device("cuda")) * 2 - 1
        11674  |        boundaries = torch.tensor([-0.9, -0.8, 0.1, 0.2, 0.5, 0.9])
        11675  |
        11676  |        self.common(fn, (inp, boundaries), check_lowp=False)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu()` function should not hardcode .cuda() call

        11700  |            self.assertEqual(ref, res)
        11701  |
        11702  |            for offset2 in (0, 1, 2, 3, 4):
    >>> 11703  |                base2 = torch.randn(
        11704  |                    64 * 64 + 64, dtype=torch.float32, device=self.device
        11705  |                ).cuda()
        11706  |                inp2 = torch.as_strided(base2, (64, 64), (64, 1), offset2)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu()` function should not hardcode torch.device('cuda')

        11725  |            return x.sin() + x.cos()
        11726  |
        11727  |        base = torch.randn(
    >>> 11728  |            64 * 64 + 64, dtype=torch.float32, device=torch.device("cuda")
        11729  |        )
        11730  |
        11731  |        inp1 = torch.as_strided(base, (32, 32), (32, 1), 4)
```


[ghstack-poisoned]

etaf committed:
Update on "[Linter] Add linter to detect device-bias hard code in test cases."


Since XPU does not gate community pull requests, we’ve observed that contributors often hardcode "cuda" in functions decorated with requires_gpu() when adding new test cases. This causes the tests to fail on XPU and breaks XPU CI.
This PR adds a linter to detect such issues automatically. An example is shown below.

```
  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode device='cuda'

        11670  |                .contiguous()
        11671  |            )
        11672  |
    >>> 11673  |        inp = torch.rand((64, 64), device="cuda") * 2 - 1
        11674  |        boundaries = torch.tensor([-0.9, -0.8, 0.1, 0.2, 0.5, 0.9])
        11675  |
        11676  |        self.common(fn, (inp, boundaries), check_lowp=False)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode .cuda() call

        11700  |            self.assertEqual(ref, res)
        11701  |
        11702  |            for offset2 in (0, 1, 2, 3, 4):
    >>> 11703  |                base2 = torch.randn(64 * 64 + 64, dtype=torch.float32).cuda()
        11704  |                inp2 = torch.as_strided(base2, (64, 64), (64, 1), offset2)
        11705  |                ref2 = fn(inp2)
        11706  |                res2 = fn_c(inp2)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode torch.device('cuda:0')

        11723  |            return x.sin() + x.cos()
        11724  |
        11725  |        base = torch.randn(
    >>> 11726  |            64 * 64 + 64, dtype=torch.float32, device=torch.device("cuda:0")
        11727  |        )
        11728  |
        11729  |        inp1 = torch.as_strided(base, (32, 32), (32, 1), 4)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode .to('cuda') call

        11771  |            torch.manual_seed(42)
        11772  |            base = torch.randn(64 * 64 + 64, dtype=torch.float32, device=self.device)
        11773  |            torch.manual_seed(42)
    >>> 11774  |            base_ref = torch.randn(64 * 64 + 64, dtype=torch.float32).to("cuda")
        11775  |
        11776  |            inp = torch.as_strided(base, size, stride, offset)
        11777  |            inp_ref = torch.as_strided(base_ref, size, stride, offset)
```


[ghstack-poisoned]

EikanWang commented on tools/linter/adapters/test_device_bias_linter.py:
```diff
@@ -0,0 +1,155 @@
+        if has_requires_gpu:
+            msg_prefix = "`@requires_gpu` function should not hardcode"
+            for subnode in ast.walk(node):
+                if isinstance(subnode, ast.keyword):
```
Pls support the case as follow as it is a normal case.

```python
if GPU_TYPE == "CUDA":
    # CUDA-specific code
```

etaf replied:
> Pls support the case as follow as it is a normal case.
> 
> ```python
> if GPU_TYPE == "CUDA":
>     # CUDA-specific code
> ```

We can't say with 100% certainty that there is a problem with this. It's ok to write such code.

EikanWang commented on tools/linter/adapters/test_device_bias_linter.py:
```diff
@@ -0,0 +1,155 @@
+                                    val,
+                                    f"{msg_prefix} torch.device('{val.args[0].value}')",
+                                )
+                if isinstance(subnode, ast.Call) and isinstance(
```
ditto

EikanWang reviewed (commented):
Overall, the idea looks good to me. @etaf , could you pls. scan the pytorch cases first and list all the cases that conflict with this linter?

etaf commented on .lintrunner.toml:
```diff
@@ -1729,3 +1729,15 @@ include_patterns = [
+    '--',
+    '@{{PATHSFILE}}',
+]
+include_patterns = [
+    'test/**/test_*.py',
+]
```
The linter only check functions decorated with `@requires_gpu` which means support both CUDA and XPU.

etaf commented:
> Overall, the idea looks good to me. @etaf , could you pls. scan the pytorch cases first and list all the cases that conflict with this linter?

I've scanned all the files before submitting this PR.

EikanWang commented on tools/linter/adapters/test_device_bias_linter.py:
```diff
@@ -0,0 +1,155 @@
+
+def main() -> None:
+    parser = argparse.ArgumentParser(
+        description="test device bias code",
```
`Detect Python functions decorated with [require_gpu] may potentially break support for other GPU devices.`

etaf replied:
> `Detect Python functions decorated with [require_gpu] may potentially break support for other GPU devices.`

Then it should be decorated with requires_cuda instead of requires_gpu.

EikanWang replied:
nit: my comment is to refine the description.

EikanWang commented on tools/linter/adapters/test_device_bias_linter.py:
```diff
@@ -0,0 +1,155 @@
+                                and isinstance(val.args[0], ast.Constant)
+                                and DEVICE_BIAS in val.args[0].value
+                            ):
+                                self.record(
```
The code block is deeply nested. It would be helpful to simplify or refactor it to improve clarity.

etaf replied:
> The code block is deeply nested. It would be helpful to simplify or refactor it to improve clarity.

Refined, please have a review.

etaf committed:
Update on "[Linter] Add linter to detect device-bias hard code in test cases."


Since XPU does not gate community pull requests, we’ve observed that contributors often hardcode "cuda" in functions decorated with requires_gpu() when adding new test cases. This causes the tests to fail on XPU and breaks XPU CI.
This PR adds a linter to detect such issues automatically. An example is shown below.

```
  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode device='cuda'

        11670  |                .contiguous()
        11671  |            )
        11672  |
    >>> 11673  |        inp = torch.rand((64, 64), device="cuda") * 2 - 1
        11674  |        boundaries = torch.tensor([-0.9, -0.8, 0.1, 0.2, 0.5, 0.9])
        11675  |
        11676  |        self.common(fn, (inp, boundaries), check_lowp=False)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode .cuda() call

        11700  |            self.assertEqual(ref, res)
        11701  |
        11702  |            for offset2 in (0, 1, 2, 3, 4):
    >>> 11703  |                base2 = torch.randn(64 * 64 + 64, dtype=torch.float32).cuda()
        11704  |                inp2 = torch.as_strided(base2, (64, 64), (64, 1), offset2)
        11705  |                ref2 = fn(inp2)
        11706  |                res2 = fn_c(inp2)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode torch.device('cuda:0')

        11723  |            return x.sin() + x.cos()
        11724  |
        11725  |        base = torch.randn(
    >>> 11726  |            64 * 64 + 64, dtype=torch.float32, device=torch.device("cuda:0")
        11727  |        )
        11728  |
        11729  |        inp1 = torch.as_strided(base, (32, 32), (32, 1), 4)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode .to('cuda') call

        11771  |            torch.manual_seed(42)
        11772  |            base = torch.randn(64 * 64 + 64, dtype=torch.float32, device=self.device)
        11773  |            torch.manual_seed(42)
    >>> 11774  |            base_ref = torch.randn(64 * 64 + 64, dtype=torch.float32).to("cuda")
        11775  |
        11776  |            inp = torch.as_strided(base, size, stride, offset)
        11777  |            inp_ref = torch.as_strided(base_ref, size, stride, offset)
```


[ghstack-poisoned]

etaf committed:
Update on "[Linter] Add linter to detect device-bias hard code in test cases."


Since XPU does not gate community pull requests, we’ve observed that contributors often hardcode "cuda" in functions decorated with requires_gpu() when adding new test cases. This causes the tests to fail on XPU and breaks XPU CI.
This PR adds a linter to detect such issues automatically. An example is shown below.

```
  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode device='cuda'

        11670  |                .contiguous()
        11671  |            )
        11672  |
    >>> 11673  |        inp = torch.rand((64, 64), device="cuda") * 2 - 1
        11674  |        boundaries = torch.tensor([-0.9, -0.8, 0.1, 0.2, 0.5, 0.9])
        11675  |
        11676  |        self.common(fn, (inp, boundaries), check_lowp=False)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode .cuda() call

        11700  |            self.assertEqual(ref, res)
        11701  |
        11702  |            for offset2 in (0, 1, 2, 3, 4):
    >>> 11703  |                base2 = torch.randn(64 * 64 + 64, dtype=torch.float32).cuda()
        11704  |                inp2 = torch.as_strided(base2, (64, 64), (64, 1), offset2)
        11705  |                ref2 = fn(inp2)
        11706  |                res2 = fn_c(inp2)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode torch.device('cuda:0')

        11723  |            return x.sin() + x.cos()
        11724  |
        11725  |        base = torch.randn(
    >>> 11726  |            64 * 64 + 64, dtype=torch.float32, device=torch.device("cuda:0")
        11727  |        )
        11728  |
        11729  |        inp1 = torch.as_strided(base, (32, 32), (32, 1), 4)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode .to('cuda') call

        11771  |            torch.manual_seed(42)
        11772  |            base = torch.randn(64 * 64 + 64, dtype=torch.float32, device=self.device)
        11773  |            torch.manual_seed(42)
    >>> 11774  |            base_ref = torch.randn(64 * 64 + 64, dtype=torch.float32).to("cuda")
        11775  |
        11776  |            inp = torch.as_strided(base, size, stride, offset)
        11777  |            inp_ref = torch.as_strided(base_ref, size, stride, offset)
```


[ghstack-poisoned]

etaf commented on tools/linter/adapters/test_device_bias_linter.py:
```diff
@@ -0,0 +1,155 @@
+            for d in node.decorator_list
+        )
+        if has_requires_gpu:
+            msg_prefix = "`@requires_gpu` function should not hardcode"
```
> Besides the error message, we can add the suggestion here as contributors may have no idea to use `GPY_TYPE`.

Done.

etaf commented on tools/linter/adapters/test_device_bias_linter.py:
```diff
@@ -0,0 +1,155 @@
+    description: str | None
+
+
+DEVICE_BIAS = "cuda"
```
> I have to admit that the device-bias cases are not dominated to `CUDA`. However, `DEVICE_BIAS` does not mean `CUDA` only. It should include other devices and should be extended easily.

Refined, thanks.

etaf committed:
Update on "[Linter] Add linter to detect device-bias hard code in test cases."


Since XPU does not gate community pull requests, we’ve observed that contributors often hardcode "cuda" in functions decorated with requires_gpu() when adding new test cases. This causes the tests to fail on XPU and breaks XPU CI.
This PR adds a linter to detect such issues automatically. An example is shown below.

```
  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode device='cuda'

        11670  |                .contiguous()
        11671  |            )
        11672  |
    >>> 11673  |        inp = torch.rand((64, 64), device="cuda") * 2 - 1
        11674  |        boundaries = torch.tensor([-0.9, -0.8, 0.1, 0.2, 0.5, 0.9])
        11675  |
        11676  |        self.common(fn, (inp, boundaries), check_lowp=False)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode .cuda() call

        11700  |            self.assertEqual(ref, res)
        11701  |
        11702  |            for offset2 in (0, 1, 2, 3, 4):
    >>> 11703  |                base2 = torch.randn(64 * 64 + 64, dtype=torch.float32).cuda()
        11704  |                inp2 = torch.as_strided(base2, (64, 64), (64, 1), offset2)
        11705  |                ref2 = fn(inp2)
        11706  |                res2 = fn_c(inp2)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode torch.device('cuda:0')

        11723  |            return x.sin() + x.cos()
        11724  |
        11725  |        base = torch.randn(
    >>> 11726  |            64 * 64 + 64, dtype=torch.float32, device=torch.device("cuda:0")
        11727  |        )
        11728  |
        11729  |        inp1 = torch.as_strided(base, (32, 32), (32, 1), 4)

  Error (TEST_DEVICE_BIAS) [device-bias]
    `requires_gpu` function should not hardcode .to('cuda') call

        11771  |            torch.manual_seed(42)
        11772  |            base = torch.randn(64 * 64 + 64, dtype=torch.float32, device=self.device)
        11773  |            torch.manual_seed(42)
    >>> 11774  |            base_ref = torch.randn(64 * 64 + 64, dtype=torch.float32).to("cuda")
        11775  |
        11776  |            inp = torch.as_strided(base, size, stride, offset)
        11777  |            inp_ref = torch.as_strided(base_ref, size, stride, offset)
```


[ghstack-poisoned]

etaf commented:
@pytorchbot merge

etaf commented:
Hi, @jansel @Skylion007 @malfet：  Would anyone be available to kindly help review this PR? I’d really appreciate it!

malfet reviewed (approved):
LGTM

malfet commented:
@pytorchbot merge

etaf commented:
@pytorchbot rebase

etaf committed:
Update

[ghstack-poisoned]

etaf commented:
@pytorchbot merge
